conda_index/postgres/cache.py

```diff
@@ -0,0 +1,416 @@
+from sqlalchemy import Connection, cte, join, or_, select
+from sqlalchemy.dialects.postgresql import insert
+
+from conda_index.index.cache import BaseCondaIndexCache, clear_newline_chars
+from conda_index.index.fs import MinimalFS
+from conda_index.index.sqlitecache import (
+    ICON_PATH,
+    PATH_TO_TABLE,
+    TABLE_NO_CACHE,
+    ChangedPackage,
+    cacher,
+    pack_record,
+)
+
+from . import model
+
+log = logging.getLogger(__name__)
+
+# prevent SQL LIKE abuse
+CHANNEL_ID_PATTERN = r"^[a-zA-Z0-9]*$"
+
+# XXX convert based on streaming "blob of json's to put in store()"
+
+
+class PsqlCache(BaseCondaIndexCache):
+    def __init__(
+        self,
+        channel_root: Path | str,
+        subdir: str,
+        *,
+        fs: MinimalFS | None = None,
+        channel_url: str | None = None,
+        upstream_stage: str = "fs",
+        db_url="postgresql://conda_index_test@localhost/conda_index_test",
+    ):
+        super().__init__(
+            channel_root,
+            subdir,
+            fs=fs,
+            channel_url=channel_url,
+            upstream_stage=upstream_stage,
+        )
+        self.db_filename = self.channel_root / ".cache" / "cache.json"
+        self.db_url = db_url
+
+        # each on-disk location gets a unique (random) prefix in the shared database
+        if not self.db_filename.exists():
+            self.db_filename.parent.mkdir(parents=True)
+            self.db_filename.write_text(json.dumps({"channel_id": os.urandom(8).hex()}))
+            self.cache_is_brand_new = True
+        else:
+            self.cache_is_brand_new = False
+
+        self.channel_id = json.loads(self.db_filename.read_text())["channel_id"]
+        if not re.match(CHANNEL_ID_PATTERN, self.channel_id):
+            raise ValueError(
+                f'{self.db_filename} contains invalid channel_id="{self.channel_id}"'
+            )
+
+    def __getstate__(self):
+        """
+        Remove db connection when pickled.
+        """
+        return {k: self.__dict__[k] for k in self.__dict__ if k not in ("engine",)}
+
+    @property
+    def database_prefix(self):
+        """
+        All paths must be prefixed with this string.
+        """
+        # If recording information about the channel_root, use '_ROOT' for nice
+        # prefix searches
+        return f"{self.channel_id}/{self.subdir or '_ROOT'}/"
+
+    @cacher
+    def engine(self):
+        engine = sqlalchemy.create_engine(self.db_url, echo=False)
+        model.create(engine)
+        return engine
+
+    def convert(self, force=False):
+        """
+        Load filesystem cache into sqlite.
+        """
+        # or call model.create(engine) here?
+        log.warning(f"{self.__class__}.convert() is not implemented")
+
+    def store_fs_state(self, listdir_stat: Iterator[dict[str, Any]]):
+        """
+        Write {path, mtime, size} into database.
+        """
+        connection: Connection
+        with self.engine.begin() as connection:
```

Comment:
Would it also help to use with `contextlib.closing(begin())` https://docs.python.org/3/library/contextlib.html#contextlib.closing for the one-off methods known to be called as futures from the parent process?

Reply:
That doesn't help, but the module-level engine cache solves the issue.

We use the https://docs.python.org/3/library/concurrent.futures.html API, where it seems like there is not a good way to limit how many times the cache class is pickled and unpickled which would cause it to create a new Engine(). There is one future per filename.